Add logical operators

v0idpwn opened the pull request:
Adds `&&`, `||` and `!` logical operators.

I think there are some improvements that can be done, but I'm opening the PR so we have a better starting point for discussion.

v0idpwn committed:
Add logical operators

polvalente commented on lib/fika/env.ex:
```diff
@@ -25,7 +25,10 @@ defmodule Fika.Env do
         "kernel./(Float,Int)" => "Float",
-        "kernel./(Float,Float)" => "Float"
+        "kernel./(Float,Float)" => "Float",
+        "kernel.||(Bool,Bool)" => "Bool",
+        "kernel.&&(Bool,Bool)" => "Bool",
+        "kernel.!(Bool)" => "Bool"
```
Maybe we could think about how to improve this soon. Also, is there any reason on why `if/else` is not a function itself like in elixir?

v0idpwn replied:
Yeah, we need a proper stdlib or something like that?

emilsoman replied:
Yeah, this is an ugly hack. This will go away as soon after we focus our attention on the compiler state - there's work planned here. Btw, in Elixir, if-else is a macro, not a function, and that expands to a case expression. Fika also converts this into a case exp in ErlTranslate.

polvalente commented on test/fika/erl_translate_test.exs:
```diff
@@ -67,6 +67,15 @@ defmodule Fika.ErlTranslateTest do
     assert result == forms
   end
 
+  test "logical operators" do
+    str = "(true || !false) && true"
+
+    ast = Parser.expression!(str)
+
+    assert {:op, 1, :and, {:op, 1, :or, {:atom, 1, true}, {:op, 1, :not, {:atom, 1, false}}},
+            {:atom, 1, true}} = ErlTranslate.translate_expression(ast)
+  end
```
I'm curious on what happens if we use `(:true || :false)`, for instance

v0idpwn replied:
As you can see in this test, in the end :boolean is cast to :atom, so I think that if something can go wrong, it will be in the type checker

polvalente replied:
yeah, at erlang level booleans don't exist

polvalente commented on test/fika/type_checker_test.exs:
```diff
@@ -40,6 +40,24 @@ defmodule Fika.TypeCheckerTest do
 
+  test "infer type of logical expressions" do
+    # and
+    str = "true && false"
```
what type is inferred by `:true && false` and `:true || false`?

polvalente commented on lib/fika/env.ex:
```diff
@@ -25,7 +25,10 @@ defmodule Fika.Env do
         "kernel./(Float,Int)" => "Float",
-        "kernel./(Float,Float)" => "Float"
+        "kernel./(Float,Float)" => "Float",
+        "kernel.||(Bool,Bool)" => "Bool",
+        "kernel.&&(Bool,Bool)" => "Bool",
```
I think we settled on `|` and `&`

sreecodeslayer commented on test/fika/parser_test.exs:
```diff
@@ -696,4 +696,51 @@ defmodule Fika.ParserTest do
+
+  describe "logic operators" do
+    test "supports consts" do
+      assert {:boolean, _line, true} = Parser.expression!("true")
```
We already have this under boolean. Do we need this here?

v0idpwn replied:
I was doing my tests and fogot, gotta remove

sreecodeslayer commented on test/fika/parser_test.exs:
```diff
@@ -696,4 +696,51 @@ defmodule Fika.ParserTest do
+      assert {:call, {:!, _}, [{:boolean, _, true}], kernel} = Parser.expression!("!true")
+    end
+
+    test "supports negation with more complex" do
```
complex expressions?

emilsoman commented on lib/fika/parser.ex:
```diff
@@ -254,10 +254,16 @@ defmodule Fika.Parser do
       literal_exps,
-      non_literal_exps
+      non_literal_exps,
+      negation
```
Can this go inside `non_literal_exps`right at the end?

emilsoman replied:
Also, we optimized the parser so a lot of things have been moved around. After you fetch the latest changes, please let me know if you need help understanding any change.

emilsoman commented on lib/fika/parser_helper.ex:
```diff
@@ -31,10 +31,20 @@ defmodule Fika.ParserHelper do
+      when bin_op in ["||", "&&"] do
+    new_left = {:call, {String.to_atom(bin_op), line}, [left, right], :kernel}
+    do_to_ast({[new_left | rest], line}, :exp_bin_op)
+  end
```
Looks like this is same as the one for +, -, *, / ? How about we add the new operators to the existing guard clause?

emilsoman commented on lib/fika/parser_helper.ex:
```diff
@@ -135,6 +145,10 @@ defmodule Fika.ParserHelper do
 
+  def do_to_ast({[{:identifier, line, value}], line}, :atom) when value in [true, false] do
+    {:boolean, line, value}
+  end
```
Why do we do this?

polvalente replied:
I suggested this so that our intermediate representation doesn't represent `:true` and `:false` differently.
Altough at the Erlang level they will be converted back to atoms, I think that some problems could arise in the type checker if we didn't have this clause

emilsoman replied:
Atoms - `:true` and `:false` in Fika are of the types `:true` and `:false`. Shouldn't they be represented differently from boolean true/false in our AST?

polvalente replied:
I disagree, because  at the BEAM level they will both be represented as atoms. Having them be different types would also make interop with Erlang and other BEAM languages much more difficult. I think that they should be booleans because of this. 

We could also go as far as disallowing `:true` and `:false` (and maybe `:nil` in the future?) as atoms altogether. This would also resolve this problem to me. In this case, the fix is not pertinent to this PR.

polvalente replied:
[This line](https://github.com/fika-lang/fika/pull/16/files#diff-cb8f4714e7283bea2cc1d4f70a924bddR99) shows that they are indeed atoms at the lowest level

emilsoman commented on test/fika/parser_test.exs:
```diff
@@ -696,4 +696,51 @@ defmodule Fika.ParserTest do
+               :kernel
+             } = result
+    end
+  end
```
Will it be possible to add some tests that separately test the operator precedence and associativity? See https://hexdocs.pm/elixir/master/operators.html for reference on precedence and associativity rules to follow.

emilsoman commented:
@v0idpwn as per the latest discussion in discord, we have agreed that for now, we'll keep :true and :false as atoms in our AST. Going forward, we may decide based on feedback from actual usage, to remove boolean keywords and use :true and :false as bools.

Documenting @polvalente 's concern:
> For anyone either coming from the BEAM or going from Fika to the lower levels, it would be really hard to differentiate both (as it is being for me).
...
what would lead someone to use :true and :false instead of the respective booleans?

v0idpwn committed:
review fixes

v0idpwn committed:
merge remote and adapt to changes

emilsoman committed:
Merge branch 'main' into feat/logical-operations

v0idpwn commented on lib/fika/parser/non_literal_exps.ex:
```diff
@@ -72,12 +72,18 @@ defmodule Fika.Parser.NonLiteralExps do
+    string("!")
+    |> parsec(:term)
+    |> Helper.to_ast(:negation)
+
```
@emilsoman I didn't manage to compile because of this `|> parsec(:term)` here.

emilsoman committed:
Fix compile error and pass tests

emilsoman committed:
Fix formatting

emilsoman committed:
Merge upstream/main

emilsoman committed:
Use &, | for logical operators

This commit adds the following rule to our syntax:
Expressions with binary operators can be broken up into multiple lines
only if the newline appears after the operator.

Valid "single" expressions:

```
x = 1 +
  2

x = foo &
  bar
```

Valid, but two expressions:

```
x = foo
 &bar
```

Invalid:

```
x = foo
  | bar
```
